fe/fe-core/src/main/java/com/starrocks/catalog/CatalogRecycleBin.java

```diff
@@ -256,40 +258,42 @@ public synchronized void replayEraseDatabase(long dbId) {
-            RecycleTableInfo tableInfo = entry.getValue();
-            Table table = tableInfo.getTable();
-            long tableId = table.getId();
-
-            if (isExpire(tableId, currentTimeMs)) {
-                if (table.getType() == TableType.OLAP) {
-                    Catalog.getCurrentCatalog().onEraseOlapTable((OlapTable) table, false);
-                }
+        for (Map<Long, RecycleTableInfo> tableEntry : idToTable.rowMap().values()) {
+            Iterator<Map.Entry<Long, RecycleTableInfo>> tableIter = tableEntry.entrySet().iterator();
+            List<Long> tableIdList = Lists.newArrayList();
+            int currentEraseOpCnt = 0;
+            while (tableIter.hasNext()) {
+                Map.Entry<Long, RecycleTableInfo> entry = tableIter.next();
+                RecycleTableInfo tableInfo = entry.getValue();
+                Table table = tableInfo.getTable();
+                long tableId = table.getId();
+
+                if (isExpire(tableId, currentTimeMs)) {
+                    if (table.getType() == TableType.OLAP) {
+                        Catalog.getCurrentCatalog().onEraseOlapTable((OlapTable) table, false);
+                    }
 
-                // erase table
-                tableIter.remove();
-                idToRecycleTime.remove(tableId);
-                tableIdList.add(tableId);
-                // log
-                LOG.info("erase table[{}-{}] in memory finished", tableId, table.getName());
-                currentEraseOpCnt++;
-                if (currentEraseOpCnt >= MAX_ERASE_OPERATIONS_PER_CYCLE) {
-                    break;
+                    // erase table
+                    tableIter.remove();
+                    idToRecycleTime.remove(tableId);
+                    tableIdList.add(tableId);
+                    // log
+                    LOG.info("erase table[{}-{}] in memory finished", tableId, table.getName());
+                    currentEraseOpCnt++;
+                    if (currentEraseOpCnt >= MAX_ERASE_OPERATIONS_PER_CYCLE) {
+                        break;
+                    }
                 }
+            } // end for tables
+            if (!tableIdList.isEmpty()) {
+                Catalog.getCurrentCatalog().getEditLog().logEraseMultiTables(tableIdList);
+                LOG.info("multi erase write log finished. erased {} table(s)", currentEraseOpCnt);
             }
-        } // end for tables
-        if (!tableIdList.isEmpty()) {
-            Catalog.getCurrentCatalog().getEditLog().logEraseMultiTables(tableIdList);
-            LOG.info("multi erase write log finished. erased {} table(s)", currentEraseOpCnt);
         }
     }
 
     private synchronized void eraseTableWithSameName(long dbId, String tableName) {
-        Iterator<Map.Entry<Long, RecycleTableInfo>> iterator = idToTable.entrySet().iterator();
+        Iterator<Map.Entry<Long, RecycleTableInfo>> iterator = idToTable.row(dbId).entrySet().iterator();
```

Comment:
Why does the function will hold the lock too long?

Reply:
I think it's caused by too many elements in idToTable, but I can't think of other reasons, and it doesn't reproduce locally.

Reply:
Database and partition also have the same problem.